### PR TITLE
Add option to specify how to lock the helper to the container

### LINKED ISSUE #21: Add support for having the helper stay in the list's bounds
Currently only it's half is inside the list.

I can submit a PR if needed.

### PR DESCRIPTION
**Pull request by zaygraveyard**

### CODE REVIEW

**Comment by clauderic** [comment]
Thanks for this. What do you think of providing a percentage value instead of a string? So inside might be 0%, outside might be 100% and in the middle (default) could be 50% of the element's height. As well, I suppose you could also provide a fixed number

**Comment by clauderic** [comment]
Also, have you checked whether this affects auto-scrolling? I have a feeling that bit of code may need tweaking as well

**Comment by zaygraveyard** [comment]
> have you checked whether this affects auto-scrolling? I have a feeling that bit of code may need tweaking as well

My first take on adding this feature was to change the `minTranslate` and `maxTranslate`, but then I saw that they were used in the auto-scrolling.
So I decided to leave the `minTranslate` and `maxTranslate`, and only update the `updatePosition()` method.
Normally this shouldn't effect the auto-scrolling, I tested a simple vertical list and it works as before.

> What do you think of providing a percentage value instead of a string?

Good idea, but in that case I don't think the `lockToContainerEdges` is the correct prop to use.
We can maybe add a `lockOffset` that takes a percentage or a px value.
Or a `lockOffsetStart` and `lockOffsetEnd` for specifying asymmetric offsets.

Any thoughts?

**Comment by clauderic** [comment]
Ah, fantastic. And yeah, `lockOffset` sounds good. If you feel strongly about supporting asymetric offsets, I think it should just be an array of two params provided to `lockOffset`

**Comment by zaygraveyard** [comment]
Great, I'm on it

**Comment by zaygraveyard** [comment]
PR updated

File: src/SortableContainer/index.js [hunk 1]
```diff
@@ -229,11 +285,17 @@ export default function SortableContainer(WrappedComponent, config = {withRef: f
...
-				if (translate[axis] < this.minTranslate) {
-					translate[axis] = this.minTranslate;
-				} else if (translate[axis] > this.maxTranslate) {
-					translate[axis] = this.maxTranslate;
-				}
+				const [minLockOffset, maxLockOffset] = this.getLockPixelOffsets();
+				const minOffset = this.dimension / 2 - minLockOffset;
+				const maxOffset = this.dimension / 2 - maxLockOffset;
+
+				translate[axis] = Math.max(
```
Comment (clauderic) [inline src/SortableContainer/index.js]: Great work! Only comment would be that this part of the code is a little obscure, I'd have a hard time wrapping my head around it if I had to debug this

**Comment by clauderic** [comment]
Would you mind updating the README.md accordingly? :)

File: src/SortableContainer/index.js (hunk 1 above)
Comment (zaygraveyard) [inline src/SortableContainer/index.js]: Do you prefer if replace it with a `limit(min, max, value)` or an `if .. else if`?

**Comment by zaygraveyard** [comment]
> Would you mind updating the README.md accordingly?

Sure no problem ^^

File: src/SortableContainer/index.js (hunk 1 above)
Comment (clauderic) [inline src/SortableContainer/index.js]: I'm fine with either, I think both would be a lot more legible and easy to understand.

**Comment by zaygraveyard** [comment]
PR updated

File: src/SortableContainer/index.js [hunk 2]
```diff
@@ -229,11 +285,15 @@ export default function SortableContainer(WrappedComponent, config = {withRef: f
...
-				}
+				const [minLockOffset, maxLockOffset] = this.getLockPixelOffsets();
+				const minOffset = this.dimension / 2 - minLockOffset;
+				const maxOffset = this.dimension / 2 - maxLockOffset;
+
+				translate[axis] = limit(
+					this.minTranslate + minOffset,
+					this.maxTranslate - maxOffset,
+					translate[axis]
+				);
```
Comment (clauderic) [inline src/SortableContainer/index.js]: 👍 Hell of a lot more legible

File: README.md [hunk 3]
```diff
@@ -102,6 +102,22 @@ More code examples are available [here](https://github.com/clauderic/react-sorta
 | useWindowAsScrollContainer | Boolean  | `false` | If you want, you can set the `window` as the scrolling container                                                                                                                                                             |
 | hideSortableGhost          | Boolean  | `true`  | Whether to auto-hide the ghost element. By default, as a convenience, React Sortable List will automatically hide the element that is currently being sorted. Set this to false if you would like to apply your own styling. |
 | lockToContainerEdges       | Boolean  | `false` | You can lock movement of the sortable element to it's parent `SortableContainer`                                                                                                                                             |
+| lockOffset                 | `OffsetValue`\* \| [`OffsetValue`\*, `OffsetValue`\*] | `"50%"` | The offsets to use when locking the movement of the sortable element. **See below for more info**. *This property has not effect of `lockToContainerEdges` is `false`.*         |
+
```
Comment (clauderic) [inline README.md]: How about something along the lines of:
> When `lockToContainerEdges` is set to `true`, this controls the offset distance between the sortable helper and the top and bottom edges of it's parent `SortableContainer`. Percentage values are relative to the height of the item currently being sorted. If you wish to specify different behaviours for locking to the *top* of the container vs the *bottom*, you may also pass in an `array` (For example: `["0%", "100%"]`).

File: README.md [hunk 4]
```diff
@@ -102,6 +102,22 @@ More code examples are available [here](https://github.com/clauderic/react-sorta
...
+A single value is equivalent to a pair that same value (`"50%"` <=> `["50%", "50%"]`).
+
+```
+ +-------------------------------+        !   Sortable List                 !
+ |      lockOffset[0] = "50%" ↕  |        !+-------------------------------+!
+=|= Item 1 ========================       !|                               |!
+!|                               |!       =|= Item 5 ========================
+!+-------------------------------+!        |      lockOffset[1] = "50%" ↕  |
+!   Sortable List                 !        +-------------------------------+
+```
```
Comment (clauderic) [inline README.md]: I would probably ditch this section, I don't really understand the drawings 😂

Comment (zaygraveyard) [inline README.md]: I wanted to put an image but I didn't find a way to do it without adding the image to this repo.
But yeah I'll remove this section :smile:

File: README.md (hunk 3 above)
Comment (zaygraveyard) [inline README.md]: Much better, thank you.

**Comment by zaygraveyard** [comment]
PR updated

File: README.md (hunk 4 above)
Comment (clauderic) [inline README.md]: You can always create a `/.github/` folder

Comment (zaygraveyard) [inline README.md]: Ok, and what do you think of these?
lockoffsetend
lockoffsetstart

Comment (clauderic) [inline README.md]: Hmm I think I'm going to make some gifs for it later, I wouldn't worry about it for the time being. Besides, there's nothing like a demo to illustrate something like this, so I'll be updating the gh-pages examples

**Comment by clauderic** [comment]
Thanks for your hard work on this man! :)

**Comment by zaygraveyard** [comment]
Happy to help :smile:

**Pull request merged**